### PR TITLE
Attr dict type check

### PR DESCRIPTION
**Pull request by d-chambers**
### What does this PR do?

This PR enables obspy.core.util.AttribDict and subclasses to perform basic type checks when assigning values to names defined in the _types dict, which stores the name of the attribute as a key and its type as the value (tuples of types also work). If the correct type was not passed a warning will be issued and the value will be cast into the stated type (or first type if tuple is used).

It is then applied on obspy.core.trace.Stats to ensure a string or None is passed to the network, station, location, and channel codes. 

For example:

```python 
from obspy.core.trace import Stats
stats = Stats()

stats.network = 'UU'  # this works fine just as before
stats.station = 1  # this issues a warning that station must be a str and casts value to '1'
stats.location = None  # also works fine
```

### Why was it initiated?  Any relevant Issues?

See: https://github.com/obspy/obspy/issues/1995

### PR Checklist
- [ ] This PR is not directly related to an existing issue (which has no PR yet).
- [X] All tests still pass.
- [X] Any new features or fixed regressions are be covered via new tests.
- [X] Any new or changed features have are fully documented.
- [X] Significant changes have been added to `CHANGELOG.txt` .
- [X] First time contributors have added your name to `CONTRIBUTORS.txt` .

### Additional Suggestions:

Additionally, I suggest the following changes to the AttribDict class:

1) Replace all lists that exist solely for membership checks with sets or tuples. This adds the advantage of more efficient look-ups or immutability (either one seems to be an advantage over a list). 

This would effect `readonly`, and `do_not_warn_on`.

2) Rename all class level attributes that can affect item/attr assignment to begin with an underscore. This would send a clear message that users should not alter them. This would include `defaults`, `readonly`, `warn_on_non_default_key` and `do_not_warn_on`. An unlikely, but possible edge-case as it stands now would allow the user to do something like the following, while only accessing "public" attributes:

```python
from obspy.core.util import AttribDict
ad = AttribDict()
ad.readonly = True  # this breaks the core functionality of the AttribDict
ad.some_value = 13  # raises TypeError because bool is not iterable
```

if, however, the user had to do this:

```python
ad._readonly = True
```

it would be clear that non-public attrs were being messed with. Should I open another issue or should we discuss here? 

edit: I just remembered in python sets are not immutable, updated comment accordingly.

### CODE REVIEW

File: obspy/core/trace.py [hunk 1]
```diff
@@ -144,6 +146,15 @@ class Stats(AttribDict):
...
         'channel': '',
     }
+    # keys which need to refresh derived values
+    _refresh_keys = {'delta', 'sampling_rate', 'starttime', 'npts'}
+    # dict of required types for certain attrs
+    _types = {
+        'network': (str, type(None)),
+        'station': (str, type(None)),
+        'location': (str, type(None)),
+        'channel': (str, type(None)),
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply obspy/core/trace.py]: Wont it be nice when py2.7 is EOL and all this silly string/future business is behind us?

**Comment by d-chambers** [comment]
I wrote a test that ensures the behavior I was after in the first place: ensuring the nslc values in Stats objects do not prevent a stream from being saved as mseed. It goes something like this:

```python
    def test_casted_stats_nscl_still_writes(self):
        """
        Ensure a Stream object that has had its nslc types cast to str can
        still be written as mseed. 
        """
        st = Stream(traces=read()[0])

        # set new stats
        with warnings.catch_warnings(record=True) as w:
            warnings.simplefilter('default')
            st[0].stats.network = 1
            st[0].stats.station = 1.1
            st[0].stats.location = b'23'
            st[0].stats.channel = None

        # try writing stream to io
        bio = io.BytesIO()
        st.write(bio, 'mseed')
```
However, st.write fails because None (stats.channel) has not attribute encode. Since the default value for the nslc codes is a blank string I don't think None should be an option (not sure why I added it). I will change the none test and remove None from the supported types dict. The way I was getting at NoneType was a bit kludgey anyhow.

**Comment by d-chambers** [comment]
Any thoughts on the additional suggestions I made in the first PR post? Worth the change or too risky?

**Comment by d-chambers** [comment]
AppVeyor is failing on python 2.7 now, will look into later this week.

**Comment by krischer** [comment]
> However, st.write fails because None (stats.channel) has not attribute encode. Since the default value for the nslc codes is a blank string I don't think None should be an option (not sure why I added it). I will change the none test and remove None from the supported types dict. The way I was getting at NoneType was a bit kludgey anyhow.

I see and also did not consider this in my first review round. I second removing `None` from the supported types dict.

Otherwise this now looks fine to me once CI passes.

**Comment by megies** [comment]
I think since we already touch this.. we might want to protect all of those special attributes from accidental overwrites? Maybe use two leading underscores and add getter properties for them?

**Comment by d-chambers** [comment]
>protect all of those special attributes from accidental overwrites? Maybe use two leading underscores and add getter properties for them?
@megies, personally I dislike the double underscore name mangling thing python does. A single underscore should be enough to tell users "don't mess with this unless you know what you are doing".

It would be easy to make a set of forbidden attrs, say (_defaults, _readonly, _warn_on_non_default_key, _do_not_warn_on) then ensure an exception is raised when a user tries to set one of these attributes.

**Comment by d-chambers** [comment]
So I tried renaming some of these attributes to start with a leading underscore but so many things broke when I ran the tests (>50). I am starting to think it may not be feasible at this point to change them. Even if I get all the tests passing who knows how many user code this would break...

**Comment by krischer** [comment]
> So I tried renaming some of these attributes to start with a leading underscore but so many things broke when I ran the tests (>50). I am starting to think it may not be feasible at this point to change them. Even if I get all the tests passing who knows how many user code this would break...

I agree - if its too hard or invasive best don't change it in this PR here.

File: obspy/core/trace.py [hunk 2]
```diff
@@ -144,6 +146,22 @@ class Stats(AttribDict):
...
     }
+    # keys which need to refresh derived values
+    _refresh_keys = {'delta', 'sampling_rate', 'starttime', 'npts'}
+    # dict of required types for certain attrs
+    # _types = {
+    #     'network': native_str,
+    #     'station': native_str,
+    #     'location': native_str,
+    #     'channel': native_str,
+    # }
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply obspy/core/trace.py]: sorry about that, not sure how that got past me...

File: obspy/core/tests/test_stats.py [hunk 3]
```diff
@@ -240,6 +243,82 @@ def test_delta_zero(self):
...
+        """
+        stats = Stats()
+        for val in self.nslc:
+
+            with warnings.catch_warnings(record=True):
+                # warnings.simplefilter('ignore')
+                setattr(stats, val, None)
+
+            # Note: removed warning capture test due to unreliably capturing
+            # warnings on py27
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply obspy/core/tests/test_stats.py]: Ya, I tried to debug it but didn't have much luck. It captures the warnings for network, station, and location just not channel. It also turns out the capture warning context manager is different on python 2 and 3 (and it works for all cases on python 3). 

When I step through it with pdb to the _cast_type method I cannot enter the warnings.warn call, and yet if I issue the command from the debugger it works and the test passes; it is the strangest bug I have seen in awhile and, unfortunately, it has me stumped. 

Do you want me to push with the failing test so you can take a look or should I just skip testing "channel"?

[the remark this answers is not in the document]
Comment (krischer) [reply obspy/core/tests/test_stats.py]: Just skip it I guess. Does not seem to be worth it to investigate further.

**Comment by krischer** [comment]
Thanks a lot for responding to all the nagging of mine. I'm happy now :) If no other comments show up I'll merge this by the end of today.

**Pull request merged**